Add bundler script that uses pyinstaller.

ivanalejandro0 opened the pull request:
How to test this:
Place the `bundle_pyinstaller.sh` file in an empty folder (for ocd reasons).
Place a sumo tarball in the folder (not strictly needed to be in same folder, ocd reasons as well)

Run: `./bundle_pyinstaller.sh sumo-tarball-name.tar.gz`

The resulting standalone bundle will be named `bitmask-version.tar.gz`

Note: you can get more colorful output logs using:
`ENABLE_COLORS=1 ./bundle_pyinstaller.sh sumo-tarball-name.tar.gz`

ivanalejandro0 committed:
Add pyinstaller bundler script.

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+#!/bin/bash
+######################################################################
+# repo-versions.sh
```
change name

ivanalejandro0 replied:
fixed

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+install_requirements
+install_sumo $sumo_path
+tweak_zope
+run_pyinstaller $bitmask_spec
```
maybe it would be useful to have passed flags to execute each one of these "targets" ?
although maybe we end up reinventing gnu-make, hehe.

ivanalejandro0 replied:
You mean like we do for `bootstrap_develop.sh:run` -> https://github.com/leapcode/bitmask_client/blob/develop/pkg/scripts/bootstrap_develop.sh#L231 ?
Yes, if we have an use case we could add that.

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+    echo "${cc_green}Status: $status...${cc_normal}"
+    set -x  # show commands
+
+    virtualenv $VENV_DIR && source $VENV_DIR/bin/activate
```
have you considered switching to venvwrapper instead? like, workon && create if fails

ivanalejandro0 replied:
I love virtualenvwrapper but, on scripts, I prefer to handle virtualenvs directly since the ammounts/complexity of steps is kind of the same, that way we don't depend on having an extra package installed and we keep the venv folder along the other stuff created for the script to run.

Why should we use venvwrapper on scripts?

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+    echo "${cc_green}Status: $status done.${cc_normal}"
+}
+
+install_pyinstaller(){
```
now that I think about it, it would make sense to check if we already have it in the venv, and install it if not.
or, probably a better idea::
- clone if dir not found in parent folder or whatever
- pull develop
- python setup.py develop instead.

ivanalejandro0 replied:
fixed

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+
+create_venv() {
+    status="creating virtualenv"
+    echo "${cc_green}Status: $status...${cc_normal}"
```
very cool, colors! :)
although we probably have to introduce a flag to disable them so that the stdio is cleaner in the bbot output

ivanalejandro0 replied:
yeah! I just copy&pasted from other scripts with colors and stuff, I didn't really thought of buildbot. I'll add a flag/env for this.
maybe `ENABLE_COLORS=1 ./bundle_pyinstaller.sh` or something.

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+        cd $REPOS_ROOT/$package
+        echo "${cc_yellow}Status: installing requirements for: $package...${cc_normal}"
+        ./pkg/pip_install_requirements.sh
+    done
```
can use the make install_base_deps I've added to bitmask makefile

ivanalejandro0 replied:
fixed

kalikaneko commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+    # - remove the zope* eggs in site-packages
+    # - remove any refs to these eggs in easy-install.pth ?? (unsure about this, talking from the top of my head) 
+
+    SITE_PACKAGES=`python -c "from distutils.sysconfig import get_python_lib; print(get_python_lib())"`
```
I guess you can factor out this command with the above use and just execute in a subshell in here

ivanalejandro0 replied:
fixed.

ivanalejandro0 commented on bundle_pyinstaller.sh:
```diff
@@ -0,0 +1,194 @@
+
+    for repo in $REPOSITORIES; do
+        echo "${cc_yellow}Status: cloning: $repo...${cc_normal}"
+        [ ! -d $repo ] && git clone -b develop $src/$repo --depth 1
```
if the repo dir exist, should I update the repo?
something like: `git fetch origin && git reset --hard develop`

I don't know how disposable will be this environment...

kalikaneko replied:
hmm I think it's a good idea to update it.
My idea is that was that we could reuse the pyinst environment for builds (we'll be exercising re-creating of venvs in other builders)

parmegv commented:
Guys, in buildbot@freenode they told me that, in scripts, it's preferable to directly use virtualenv/bin/pip, virtualenv/bin/python... without virtualenv/bin/activate. The reason is that activate is meant to be used interactively. I've implemented bitmask-buildbot following that guideline, with success. What do you think?

parmegv commented:
It would be great if there were a script to download the osrandom_engine.[ch].

Something like

```
wget -qO- https://pypi.python.org/packages/source/c/cryptography/cryptography-0.9.3.tar.gz#md5=70af3fbce7683b31d173eba225f18650
| tar xz --strip-components=7 cryptography-0.9.3/src/cryptography/hazmat/bindings/openssl/src/osrandom_engine.c cryptography-0.9.3/src/cryptography/hazmat/bindings/openssl/src/osrandom_engine.h && find -name osrandom_engine.h
```

kalikaneko commented:
indeed. feel free to add it.
however, instead of adding it each time, I would check if the files are in some temporary directory..
this is only a hack around a missing pyinstaller hook, I should submit a patch to pyinstaller to gather those.

parmegv commented:
https://github.com/ivanalejandro0/bitmask_bundler/pull/1

parmegv committed:
[feat] Detect sumo top folder with tar

The script was assuming that the sumo tarball was named after the
version of bitmask it contains, but "latest" symlink generated by
buildbot doesn't follow that rule.

Let tar list the folder the tarball generates (assuming there is only
one folder, and that it's listed in first place), and follow accordingly.

- Related: #7345

ivanalejandro0 commented:
> It would be great if there were a script to download the osrandom_engine.[ch].

I've updated the script, now it gets the cryptography files from the installed package in the current virtualenv (as for u1db file)... Sorry for not seeing this before you write the downloader script.

ivanalejandro0 commented:
> Guys, in buildbot@freenode they told me that, in scripts, it's preferable to directly use virtualenv/bin/pip, virtualenv/bin/python... without virtualenv/bin/activate. The reason is that activate is meant to be used interactively. I've implemented bitmask-buildbot following that guideline, with success. What do you think?

You mean by setting this https://github.com/parmegv/buildbot-bitmask/blob/master/setup.sh#L10, right?
I've never think about this since `activate` worked always great for me. It seems like in order to not use that script we need to manually tweak the environment as the `activate` script does, right?
I'm worried about doing this manually and not considering some case (besides changing `$PATH`), I really trust on virtualenv so I'm not worried about virtualenv breaking something.
Maybe I'm missing something.

parmegv commented:
I may be missing the same thing then hehehe. The activate method has some problems (see https://gist.github.com/datagrok/2199506#problems), but I don't have a strong opinion other than a great developer suggesting me not to use activate in a script.

parmegv commented:
Oh, update your instructions in the description of the pullreq: we don't need to download osrandom_engine anymore thanks to your fix :)
